src/maps/mod.rs

```diff
@@ -269,194 +237,194 @@ impl PerfMap {
-                            ..(std::mem::size_of::<PerfEventHeader>() + std::mem::size_of::<u32>()),
-                        )
-                        // might be unnecesary but I like this being explicit
-                        .for_each(|_| {});
-
-                        buf
-                    };
-
-                    let sample = PerfEventSample { header, size, data };
-                    Ok(Some(PerfEvent::Sample(sample)))
-                }
-                perf_event_type::PERF_RECORD_LOST => {
-                    let lost = &*(buf.as_ptr() as *const PerfEventLostSamples);
-                    Ok(Some(PerfEvent::Lost(lost.count)))
-                }
-                _ => Ok(None),
-            }
-        }
+                Ok(PerfMap {
+                    name: map_name.to_owned(),
+                    base_ptr: AtomicPtr::new(base_ptr),
+                    page_count,
+                    page_size,
+                    mmap_size,
+                    cpuid,
+                    ev_fd: fd,
+                    ev_name: "".to_owned(),
+                })
+            })
+            .collect()
     }
 
     /// Reads all available events
     ///
-    /// Stops reading either on a real error (and propagates it) or on
-    /// a OxidebpfError::NoPerfData and returns `None`.
-    pub(crate) fn read_all(&self) -> impl Iterator<Item = Result<PerfEvent, OxidebpfError>> + '_ {
-        PerfEventIterator {
-            map: self,
-            is_done: false,
-        }
+    /// Stops reading if it encounters an unexpected perf event.
+    ///
+    /// When the returned iterator is dropped it internally marks the
+    /// data as "read" so the ebpf program can re-use that
+    /// data. Because of this we should process the iterator fast as
+    /// to free space for more events.
+    ///
+    /// # Safety
+    ///
+    /// This is only safe if a single iterator is running per perfmap.
+    /// This function is marked as `&self` for easiness of use and
+    /// because it is internal only but it probably should be `&mut
+    /// self`. When the iterator is dropped it internally changes data
+    /// in the mmap that the kernel manages (data_tail to be precise)
+    /// to tell it what is the last bit we read so we shouldn't have
+    /// multiple mutations at the same time.
+    pub(crate) unsafe fn read_all(
+        &self,
+    ) -> impl Iterator<Item = Result<PerfEvent, OxidebpfError>> + '_ {
+        PerfEventIterator::new(self)
     }
 }
 
 struct PerfEventIterator<'a> {
-    map: &'a PerfMap,
-    is_done: bool,
+    // modified by iterator
+    data_tail: u64,
+    data_head: u64,
+    errored: bool,
+    copy_buf: Vec<u8>, // re-usable buffer to make ring joins be contiguous
+
+    // calculated at creation
+    mmap_size: usize,
+    base: *const u8,
+    metadata: *mut PerfMem,
+
+    // technically not needed but it gives us the lifetime we need to
+    // prevent the iterator outliving the perfmap
+    _map: &'a PerfMap,
```

Comment:
If we don't actually need it, maybe a `PhantomData<&'a PerfMap>` would be better to avoid adding the size of a reference to the structure?

Reply:
Relevant Rustonomicon section: https://doc.rust-lang.org/nomicon/phantom-data.html